flink-table-store-core/src/main/java/org/apache/flink/table/store/file/data/AppendOnlyCompactManager.java

```diff
@@ -0,0 +1,131 @@
+        this.rewriter = rewriter;
+    }
+
+    @Override
+    public void submitCompaction() {
+        if (taskFuture != null) {
+            throw new IllegalStateException(
+                    "Please finish the previous compaction before submitting new one.");
+        }
+        pickCompactBefore()
+                .ifPresent(
+                        (before) -> {
+                            if (LOG.isDebugEnabled()) {
+                                LOG.debug(
+                                        "Submit compaction with files (name, level, size): "
+                                                + before.stream()
+                                                        .map(
+                                                                file ->
+                                                                        String.format(
+                                                                                "(%s, %d, %d)",
+                                                                                file.fileName(),
+                                                                                file.level(),
+                                                                                file.fileSize()))
+                                                        .collect(Collectors.joining(", ")));
+                            }
+                            taskFuture =
+                                    executor.submit(
+                                            new CompactTask() {
+                                                @Override
+                                                protected void doCompact() throws Exception {
+                                                    compactBefore.addAll(before);
+                                                    compactAfter.addAll(rewriter.rewrite(before));
+                                                }
+                                            });
+                        });
+    }
+
+    @VisibleForTesting
+    Optional<List<DataFileMeta>> pickCompactBefore() {
+        long totalFileSize = 0L;
+        int fileNum = 0;
+        LinkedList<DataFileMeta> compactBefore = new LinkedList<>();
+
+        while (!toCompact.isEmpty()) {
+            DataFileMeta file = toCompact.pollFirst();
+            compactBefore.add(file);
+            totalFileSize += file.fileSize();
+            fileNum++;
+            if ((totalFileSize >= targetFileSize && fileNum >= minFileNum)
+                    || fileNum >= maxFileNum) {
+                return Optional.of(compactBefore);
+            } else if (totalFileSize >= targetFileSize) {
+                // left pointer shift one pos to right
+                DataFileMeta removed = compactBefore.pollFirst();
+                assert removed != null;
+                totalFileSize -= removed.fileSize();
+                fileNum--;
+            }
+        }
+        for (DataFileMeta file : compactBefore) {
```

Comment:
This time, `toCompact` is empty? Assert this?

Reply:
![image](https://user-images.githubusercontent.com/55568005/177935047-492260da-8656-4fea-8b57-898e90e59495.png)

I think the assertion is optional.